### ISSUE TITLE
Error in TM Unit Test, hearkens on another potential issue

### ISSUE THREAD

**Issue by cogmission** [numenta/nupic#3294]
```temporal_memory_test.py``` [successfully creates a segment here](https://github.com/numenta/nupic/blob/master/tests/unit/nupic/research/temporal_memory_test.py#L636)  on a ```Cell``` which doesn't exist according to the parameters set on the TemporalMemory both [here](https://github.com/numenta/nupic/blob/master/tests/unit/nupic/research/temporal_memory_test.py#L618) and [here](https://github.com/numenta/nupic/blob/master/tests/unit/nupic/research/temporal_memory_test.py#L619).

According to these, there should only be at most 32 cells...

Aside from the slight oversight in the test, the question that comes to mind is, should this be prevented? I'm thinking it should...?

**Comment by rhyolight** [comment]
Good question. Possible bug? @numenta/nupic-committers Anyone else think this test looks suspect?

**Comment by natoromano** [comment]
In practice tm.connections aren't really be accessed from the outside, so this can't really prevented by the TM itself. I think (and hope) it is only creating synapses on correct cells when it is calling its connections member. 

However I think Connections objects should absolutely be allowed to create connections to out-of-range presynaptic cell, as those can be in a higher layer of the hierarchy (for apical connections) or from a different source (for distal input). In those case the cells are often re-indexed (such as in the research Extended Temporal Memory), and even without that, they can come from a layer which has many more cells.

TL;DR, No, it's not suspect a behavior. The only question is, is that really useful in that particular test...

**Comment by cogmission** [comment]
@natoromano

I am concerned about this. Here's my reasons. The TM has a Connections local field coupling the particular Connections instance to a specific instance of a given TM. That TM then has local fields: "columnDimensions" and "cellsPerColumn", further cementing the coupling of the configurations. Furthermore, mappings identifying segments and synapses are keyed in particular indexes of cells and segments, together with reverse mappings of presynaptic cells to their receiving synapses - which all reside in a particular Connections object. The problem is even though we are trending torwards more object representations (with the advent of Segment and Synapse / data objects), columns are still being derived rather than mapped - as cellIndexes / cellsPerColumn which is a local configuration.

Has there been any consideration for how scope will be applied to Connections objects if as you say there will be mappings created which cross apical and lateral/distal boundaries? I.e. Cell 81 will most likely also be redundantly mapped to another body of columns and cells too? Is there a vision for where this is headed architecturally ?  :-)

**Comment by natoromano** [comment]
In the current implementation, (distal) input "flows" through the same Connections object, starting at index numberOfCells (the first numberOfCells are for internal connections), and apicalConnections have a Connections object all for themselves (but the inputs are still reindexed so that external input and actual cells from the TM never have the same index).

**Comment by cogmission** [comment]
> but the inputs are still reindexed

I can't visualize this, but as long as there's a design for how this fits together, I'm cool with it. I just think the local test should be changed to the ```columnDimension = [ 100 ]``` specification for the local test so that anyone reading it will not be confused since it is a local unit test?

**Comment by natoromano** [comment]
Well, let's say you are training a 32-cells TM with an external input which is another 32-cells TM. If the cell 5 from the external TM fires, it will be fed in as the cell 37 to the main TM's Connections.

I have no idea was this index was used for this test, and I agree that it is confusing and should be changed if there is no particular reason for it be outside the TM's range.

**Comment by cogmission** [comment]
> Well, let's say you are training a 32-cells TM with an external input which is another 32-cells TM. If the cell 5 from the external TM fires, it will be fed in as the cell 37 to the main TM's Connections. 
 
Wow. It'll be an interesting challenge to handle all this from an OO point of view where right now indexes outside of the "known" ranges result in null pointer exceptions, which is how my attention was brought to this issue during porting.

**Comment by cogmission** [comment]
Trying to tidy up my cognitive residue. 

This boils down to two parts:
* The test should be changed so that it doesn't reference indexes that don't exist - although legal and not a inappropriate feature, it isn't relevant to the context of the test, and causes cognitive dissonance.

* The question of indexes which lie outside the configured boundaries of a TM or SP's cells/columns has been explained. It is "legal" and appropriate and is/will be a common practice.

Please tell me if I didn't boil this down succinctly: @natoromano @rhyolight ? This should be turned into an issue to make the slight "nit" fix to the test.

**Comment by rhyolight** [comment]
> This should be turned into an issue to make the slight "nit" fix to the test.

Sounds good to me... @natoromano ?

**Comment by natoromano** [comment]
Yes @cogmission @rhyolight , to me this is exactly it.

I re-read the test and really can't think of a reason why this index is used, it is indeed confusing. A "valid" (i.e. internal to the TM) index should be used.

**Comment by rhyolight** [comment]
@cogmission @natoromano Can one of you either (a) edit this ticket to represent the action work that needs to done to fix the test or (b) close this issue and create another.

Thanks!
